contrib/win32/win32compat/misc.c

```diff
@@ -2119,3 +2124,95 @@ strrstr(const char *inStr, const char *pattern)
 
 	return last;
 }
+
+int
+add_mark_of_web(const char* filename)
+{
+	if (motw_zone_id > 4) {
+		return -1;
+	}
+	char* fileStreamPath = NULL;
+	size_t fileStreamPathLen = strlen(filename) + strlen(":Zone.Identifier") + 1;
+
+	fileStreamPath = malloc(fileStreamPathLen * sizeof(char));
+
+	if (fileStreamPath == NULL) {
+		return -1;
+	}
+
+	sprintf_s(fileStreamPath, fileStreamPathLen, "%s:Zone.Identifier", filename);
+
+	int ofd, status = 0;
+	char* zoneIdentifierStr = NULL;
+	size_t zoneIndentifierLen = strlen("[ZoneTransfer]\nZoneId=") + 1 + 1;
+
+	zoneIdentifierStr = malloc(zoneIndentifierLen * sizeof(char));
+
+	if (zoneIdentifierStr == NULL) {
+		status = -1;
+		goto cleanup;
+	}
+
+	sprintf_s(zoneIdentifierStr, zoneIndentifierLen, "[ZoneTransfer]\nZoneId=%d", motw_zone_id);
+
+	// create zone identifer file stream and then write the Mark of the Web to it
+	if ((ofd = open(fileStreamPath, O_WRONLY | O_CREAT, USHRT_MAX)) == -1) {
+		status = -1;
+		goto cleanup;
+	}
+
+	if (atomicio(vwrite, ofd, zoneIdentifierStr, zoneIndentifierLen) != zoneIndentifierLen) {
+		status = -1;
+	}
+
+	if (close(ofd) == -1) {
+		status = -1;
+	}
+
+cleanup:
+	free(fileStreamPath);
+	if (zoneIdentifierStr)
+		free(zoneIdentifierStr);
+	return status;
+}
+
+/* Gets the zone identifier value based on the provided hostname, 
+and sets the global motw_zone_id variable with that value. */
+void get_zone_identifier(const char* hostname) {
+	static const CLSID CLSID_ISM =
+	{ 0x7B8A2D94, 0x0AC9, 0x11D1,
+	{ 0x89, 0x6C, 0x00, 0xC0, 0x4F, 0xB6, 0xBF, 0xC4 } };
+	static const IID IID_IISM =
+	{ 0x79EAC9EE, 0xBAF9, 0x11CE,
+	{ 0x8C, 0x82, 0x00, 0xAA, 0x00, 0x4B, 0xA9, 0x0B } };
+	IInternetSecurityManager* IISM;
+	CoInitialize(NULL);
+	HRESULT hr = CoCreateInstance(&CLSID_ISM, NULL,
+		CLSCTX_ALL, &IID_IISM, (void**)&IISM);
+	if (!(SUCCEEDED(hr) || hr == RPC_E_CHANGED_MODE)) {
```

Comment:
I believe `hr == RPC_E_CHANGED_MODE` is a success condition?
```suggestion
	if (!(SUCCEEDED(hr) && hr != RPC_E_CHANGED_MODE)) {
```

Reply:
Ah, this was supposed to be a check on the [CoInitialize](https://docs.microsoft.com/en-us/windows/win32/api/objbase/nf-objbase-coinitialize) call above (also realized I should be using [CoInitializeEx](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-coinitializeex)). 
For that - either success(hr) or RPC_E_CHANGED_MODE indicates success. Although, I don't know that RPC_E_CHANGED_MODE is required in this case, since we are not calling CoInitializeEx anywhere else?

Reply:
I don't think we call CoInitialize anywhere else in the code base, which makes sense since I believe this is the first time we want to use a COM API.  I think we should just check `hr` for success and ignore the RPC_E_CHANED_MODE code.